tauri-app/src-tauri/src/commands/fork.rs

```diff
@@ -0,0 +1,191 @@
+use alloy_dyn_abi::JsonAbiExt;
+use alloy_json_abi::Error;
+use forker::*;
+use once_cell::sync::Lazy;
+use reqwest::Client;
+use revm::primitives::Bytes;
+use std::{collections::HashMap, sync::Mutex};
+
+/// static hashmap of fork evm instances, used for caching instances between runs
+pub static FORKS: Lazy<Mutex<HashMap<String, ForkedEvm>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
+
+/// hashmap of cached error selectors    
+pub static SELECTORS: Lazy<Mutex<HashMap<Vec<u8>, Error>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
+
+#[tauri::command]
+pub async fn fork_call(
+    fork_url: &str,
+    fork_block_number: u64,
+    from_address: &[u8],
+    to_address: &[u8],
+    calldata: &[u8],
+) -> Result<Result<Bytes, String>, String> {
+    let result = {
+        // lock static FORKS
+        let mut forks = FORKS.lock().unwrap();
+
+        // build key from fork url and block number
+        let key = fork_url.to_owned() + &fork_block_number.to_string();
+
+        // fork from the provided url, if it is cached, use it, if not create it, and cache it in FORKS
+        let forked_evm = if let Some(v) = forks.get_mut(&key) {
+            v
+        } else {
+            let new_forked_evm = ForkedEvm::new(
+                None,
+                fork_url.to_owned(),
+                Some(fork_block_number),
+                200000u64,
+            );
+            forks.insert(key.clone(), new_forked_evm);
+            forks.get_mut(&key).unwrap()
+        };
+
+        // call a contract read-only
+        forked_evm
+            .call(from_address, to_address, calldata)
+            .map_err(|e| e.to_string())?
+    };
+
+    if result.reverted {
+        // decode result bytes to error selectors if it was a revert
+        Err(decode_error(&result.result).await?)
+    } else {
+        Ok(Ok(result.result))
+    }
+    // Ok(Ok(()))
+}
+
+/// decodes an error returned from calling a contract by searching its selector in registry
+async fn decode_error(error_data: &[u8]) -> Result<String, String> {
+    let (selector_hash_bytes, args_data) = error_data.split_at(4);
+    let selector_hash = alloy_primitives::hex::encode_prefixed(selector_hash_bytes);
+
+    // check if selector already is cached
+    {
+        let selectors = SELECTORS.lock().unwrap();
+        if let Some(error) = selectors.get(selector_hash_bytes) {
+            if let Ok(result) = error.abi_decode_input(args_data, false) {
+                return Ok(format!("{}: {:?}", error.name, result));
+            } else {
+                return Ok("unknown error".to_owned());
```

Comment:
return an enum so the caller can handle the unknown error without string matching

Reply:
this is meant to be shown on the problem panel, that's why it is string,